File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    # dp[i][j]: (i, j)까지 가는 경로의 수
    dp = [[0] * m for _ in range(n)]
    dp[0][0] = 1  # 시작점

    # 물웅덩이를 set에 저장 (빠른 조회를 위해)
    puddles_set = {(py - 1, px - 1) for px, py in puddles}

    # 첫 번째 열 초기화 (물웅덩이 만나면 이후는 갈 수 없음)
    for x in range(n):
        if (x, 0) in puddles_set:
            break
        dp[x][0] = 1

    # 첫 번째 행 초기화 (물웅덩이 만나면 이후는 갈 수 없음)
    for y in range(m):
        if (0, y) in puddles_set:
            break
        dp[0][y] = 1

    # DP 테이블 채우기
    for i in range(1, n):
        for j in range(1, m):
            if (i, j) in puddles_set:  # 물웅덩이이면 스킵
                dp[i][j] = 0
            else:
                dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % 1000000007

    return dp[-1][-1]

File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;


class Solution {
    
    public int solution(int m, int n, int[][] puddles) {
        boolean[][] isPuddle = new boolean[n][m];
        for (int[] puddle : puddles) {
            isPuddle[puddle[1]-1][puddle[0]-1] = true;
        }
        
        
        int[][] dp = new int[n][m];
        
        for (int x = 1; x < n; x++) {
            if (isPuddle[x][0]) {
                break;
            }
            dp[x][0] = 1;
        }
        for (int y = 1; y < m; y++) {
            if (isPuddle[0][y]) {
                break;
            }
            dp[0][y] = 1;
        }
        
        for (int x = 1; x < n; x++) {
            for (int y = 1; y < m; y++) {
                if (isPuddle[x][y]) {
                    continue;
                }
                dp[x][y] = (dp[x-1][y] + dp[x][y-1]) % 1000000007;
            }
        }
        
        return dp[n-1][m-1];
    }
    
}